### PR TITLE
1.27 Backport H/2 header discard fix

### PR DESCRIPTION
**Pull request by SeanKilleen**
Commit Message:
Discard the Host header if the :authority header was received to bring Envoy into compliance with https://www.rfc-editor.org/rfc/rfc9113#section-8.3.1

This behavioral change can be reverted by setting runtime flag envoy.reloadable_features.http2_discard_host_header to false.

Risk Level: Low
Testing: Unit Tests
Docs Changes: N/A
Release Notes: Yes
Platform Specific Features: N/A
Runtime guard: envoy.reloadable_features.http2_discard_host_header
Back-ports #30005, supporting #31118

### CODE REVIEW

File: test/integration/multiplexed_integration_test.cc [hunk 1]
```diff
@@ -2129,7 +2131,86 @@ TEST_P(Http2FrameIntegrationTest, AccessLogOfWireBytesIfResponseSizeGreaterThanW
...
+  EXPECT_EQ(Http2Frame::ResponseStatus::Ok, frame.responseStatus());
+  tcp_client_->close();
+}
+
+// All HTTP/2 static headers must be before non-static headers.
+// Verify that codecs validate this.
+TEST_P(Http2FrameIntegrationTest, HostBeforeAuthorityIsRejected) {
+#ifdef ENVOY_ENABLE_UHV
+  // TODO(yanavlasov): fix this check for oghttp2 in UHV mode.
+  if (GetParam().http2_implementation == Http2Impl::Oghttp2) {
```
Comment (yanavlasov) [inline test/integration/multiplexed_integration_test.cc:2191]: To simplify this, please revert the change to the `FrameIntegrationTestParam` and the `testParams()` method and return unconditionally in this `#ifdef` block.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
Thanks for doing this backport.

/wait

File: test/integration/multiplexed_integration_test.cc (hunk 1 above)
Comment (SeanKilleen) [reply test/integration/multiplexed_integration_test.cc:2191]: Sounds good, will do now. Thanks for the feedback!

File: test/integration/multiplexed_integration_test.cc [hunk 2]
```diff
@@ -2129,7 +2129,84 @@ TEST_P(Http2FrameIntegrationTest, AccessLogOfWireBytesIfResponseSizeGreaterThanW
...
+  sendFrame(request);
+
+  waitForNextUpstreamRequest();
+  EXPECT_EQ(upstream_request_->headers().getHostValue(), "one.example.com");
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+  auto frame = readFrame();
+  EXPECT_EQ(Http2Frame::Type::Headers, frame.type());
+  EXPECT_EQ(Http2Frame::ResponseStatus::Ok, frame.responseStatus());
+  tcp_client_->close();
+}
```
Comment (phlax) [inline test/integration/multiplexed_integration_test.cc:2164]: i think these 2 tests have been added in this pr - i guess because the original pr made alterations to them

**Review by phlax — APPROVED** [review]
added tests dont seem to break anything so ill land this to include in the release

lgtm, thanks again

@SeanKilleen

File: test/integration/multiplexed_integration_test.cc (hunk 1 above)
Comment (phlax) [reply test/integration/multiplexed_integration_test.cc]: resolved, thanks
